lib/Memoizer.js

```diff
@@ -1,61 +1,117 @@
-/* globals Services */
+/* globals Services, Task */
 
 "use strict";
 const {Cu} = require("chrome");
 const simplePrefs = require("sdk/simple-prefs");
+const ss = require("sdk/simple-storage");
 Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+
+const concat = String.prototype.concat;
 
 function Memoizer() {
   this._cacheEnabled = simplePrefs.prefs["query.cache"];
-  this._memo_all = {};
+
+  if (!ss.storage.queryCache) {
+    ss.storage.queryCache = {};
+  }
   this._onPrefChange = this._onPrefChange.bind(this);
   simplePrefs.on("", this._onPrefChange);
 }
 
 Memoizer.prototype = {
 
+  /**
+   * Return a function that will wrap a given function and cache its results.
+   *
+   * @param {String} key
+   *   a key to store/retrieve cached values
+   *
+   * @param {Function} func
+   *   a function to wrap for caching
+   *
+   * @return {Promise}
+   *   A promise whose result contains data
```

Comment:
This actually returns a function that returns a Promise if I am reading it right.